awd/database.py

```diff
@@ -0,0 +1,132 @@
+from __future__ import annotations
+
+import datetime
+import logging
+from typing import Any
+
+from pynamodb.attributes import NumberAttribute, UnicodeAttribute
+from pynamodb.exceptions import DoesNotExist
+from pynamodb.models import Model
+
+from awd.config import DATE_FORMAT
+from awd.status import Status
+
+logger = logging.getLogger(__name__)
+
+
+class DoiProcessAttempt(Model):
+    """A class modeling an item in the DynamoDB table."""
+
+    class Meta:  # noqa: D106
+        table_name = "None"
+
+    doi = UnicodeAttribute(hash_key=True)
+    attempts = NumberAttribute()
+    last_modified = UnicodeAttribute()
+    status = NumberAttribute()
+
+    @classmethod
+    def add_item(cls, doi: str) -> dict[str, Any]:
+        """Add DOI item to DOI table.
+
+        Args:
+            doi: The DOI to be added to the DOI table.
+        """
+        response = cls(
+            doi=doi,
+            attempts=0,
+            last_modified=datetime.datetime.now(tz=datetime.UTC).strftime(DATE_FORMAT),
+            status=Status.UNPROCESSED.value,
+        ).save()
+        logger.debug("%s added to table", doi)
+        return response
+
+    def has_unprocessed_status(self, doi: str) -> bool:
+        """Validate that a DOI has unprocessed status in the DOI table."""
+        return DoiProcessAttempt.get(doi).status == Status.UNPROCESSED.value
+
+    def check_doi_and_add_to_table(self, doi: str) -> None:
+        """Check if DOI should be added to table.
+
+        If not present, add the DOI to the table.  Check for unprocessed status
+        and raise exception if DOI should not be retried. Increment attempts field.
+
+        Args:
+            doi: The DOI to be checked and possibly added to the DOI table.
+        """
+        try:
+            self.get(doi)
+        except DoesNotExist:
+            self.add_item(doi)
+        if not self.has_unprocessed_status(doi):
+            raise UnprocessedStatusFalseError
+        self.increment_attempts(doi)
```

Comment:
Again, easy to follow.

Given recent discussions about methods doing "one thing", I find that the docstring here is invaluable as a couple of things do happen, but it's all the same levels of abstraction.  

And when you can do something as simple as:
```python
try:
            self.get(doi)
        except DoesNotExist:
```

you really don't need a dedicated method for that!